toolbox/deps_update/main.go

```diff
@@ -0,0 +1,259 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+
+	"istio.io/test-infra/toolbox/util"
+)
+
+var (
+	repo       = flag.String("repo", "", "Update dependencies of only this repository")
+	owner      = flag.String("owner", "istio", "Github Owner or org")
+	tokenFile  = flag.String("token_file", "", "File containing Github API Access Token")
+	baseBranch = flag.String("base_branch", "master", "Branch from which the deps update commit is based")
+	githubClnt *githubClient
+)
+
+const (
+	istioDepsFile = "istio.deps"
+)
+
+// Update the commit SHA reference in a given line from dependency file
+// to the latest stable version
+// Returns the updated line
+func replaceCommit(line string, dep dependency) string {
+	idx := strings.Index(line, "\"")
+	return line[:idx] + "\"" + dep.LastStableSHA + "\","
+}
+
+// Generates an MD5 digest of the version set of the repo dependencies
+// useful in avoiding making duplicate branches of the same code change
+// Also updates dependency objects deserialized from istio.deps
+func fingerPrintAndUpdateDepSHA(repo string, deps *[]dependency) (string, error) {
+	digest, err := githubClnt.getHeadCommitSHA(repo, *baseBranch)
+	if err != nil {
+		return "", err
+	}
+	digest += *baseBranch
+	for i, dep := range *deps {
+		commitSHA, err := githubClnt.getHeadCommitSHA(dep.RepoName, dep.ProdBranch)
+		if err != nil {
+			return "", err
+		}
+		digest += commitSHA
+		(*deps)[i].LastStableSHA = commitSHA
+	}
+	return util.GetMD5Hash(digest), nil
+}
+
+// Read the exported environment variable named :key in istio.VERSION
+// and return its value
+func istioVersionGet(key string) (string, error) {
```

Comment:
It is difficult to create such a map. I can source the exported vars in istio.VERSION to the environment, but I won't know which variables are exactly the ones I just exported. To make a map, I could either read the entire env and put it into map (huge), or I could manually parse istio.VERSION line by line (file and format specific, create source of bugs once istio.VERSION changes). Neither of those solutions is as simple and maintainable as the one I have now. But do let me know if there is better way to do this.